src/abstractdataframe/reshape.jl

```diff
@@ -185,50 +185,32 @@ long = stack(wide)
 wide0 = unstack(long)
 wide1 = unstack(long, :variable, :value)
 wide2 = unstack(long, :id, :variable, :value)
+wide3 = unstack(long, [:id, :a], :variable, :value)
 ```
 Note that there are some differences between the widened results above.
 
 """
-function unstack(df::AbstractDataFrame, rowkey::Int, colkey::Int, value::Int)
```

Comment:
Can you explain why you find it natural that the `groupby` approach is more efficient than the approach used here? AFAIK it should be quite faster, especially if the original column is a `CategoricalArray` (since we don't have an optimized `groupby` implementation for `CategoricalArray` at the moment, and it's not trivial to do).

Since the code exists, I'm reluctant to remove it since we have sometimes found out that some trivial optimizations could be applied which made a dramatic difference.

Reply:
See the benchmark:
```
julia> x = DataFrame(rand(1000, 10));

julia> x[:id] = 1:1000;

julia> y = stack(x);

julia> z = copy(y);

julia> categorical!(z, :id);

julia> eltypes(y)
3-element Array{Type,1}:
 Symbol
 Float64
 Int64

julia> eltypes(z)
3-element Array{Type,1}:
 Symbol
 Float64
 CategoricalArrays.CategoricalValue{Int64,UInt32}

julia> @benchmark unstack(y)
BenchmarkTools.Trial:
  memory estimate:  2.01 MiB
  allocs estimate:  78752
  --------------
  minimum time:     7.898 ms (0.00% GC)
  median time:      8.700 ms (0.00% GC)
  mean time:        8.998 ms (3.72% GC)
  maximum time:     59.264 ms (84.15% GC)
  --------------
  samples:          556
  evals/sample:     1

julia> @benchmark unstack(y, :variable, :value)
BenchmarkTools.Trial:
  memory estimate:  1.85 MiB
  allocs estimate:  57636
  --------------
  minimum time:     6.008 ms (0.00% GC)
  median time:      6.769 ms (0.00% GC)
  mean time:        6.965 ms (3.64% GC)
  maximum time:     58.657 ms (87.87% GC)
  --------------
  samples:          717
  evals/sample:     1

julia> @benchmark unstack(z)
BenchmarkTools.Trial:
  memory estimate:  1.65 MiB
  allocs estimate:  68695
  --------------
  minimum time:     10.947 ms (0.00% GC)
  median time:      11.326 ms (0.00% GC)
  mean time:        11.677 ms (2.53% GC)
  maximum time:     64.046 ms (81.62% GC)
  --------------
  samples:          428
  evals/sample:     1

julia> @benchmark unstack(z, :variable, :value)
BenchmarkTools.Trial:
  memory estimate:  1.92 MiB
  allocs estimate:  56270
  --------------
  minimum time:     9.071 ms (0.00% GC)
  median time:      9.890 ms (0.00% GC)
  mean time:        10.179 ms (2.94% GC)
  maximum time:     65.845 ms (84.51% GC)
  --------------
  samples:          491
  evals/sample:     1
```

The difference is because of the lines:
```
i = Int(CategoricalArrays.order(refkeycol.pool)[refkeycol.refs[k]]) # more expensive
```
and
```
i = rowkey[k] # cheaper
```

Do you see any area for optimization? If single-column version is not significantly faster I would remove it as now:
1. we have a duplicate implementation of the same feature (which increases package complexity and maintenance cost);
2. the single-column version has a bug when unstacking on id-column that has missing values.

Reply:
Yeah, the second line is cheaper, but it only comes after a call to `groupby` which should take much more time. So it looks like something is going wrong with the other method. Have you tried profiling it?

I agree with you that removing duplicated methods is generally a good idea. But I don't want to do that until we are certain to understand the situation, and here it looks like something doesn't work as I would expect it to. For example, we have recently realized that building a `CategoricalArray` was much slower than it should have been. If we could fix a similar issue here, lots of places would benefit from the fix (even if in the end we remove the method).

A possibility here is that constructing a `CategoricalArray` is relatively slow since IDs are all unique, and `CategoricalArray` isn't efficient for this kind of data. Or maybe the compiler is not able to move the `CategoricalArrays.order(keycol.pool)` call outside of the loop, i.e. we should manually store the order in a variable?

Reply:
I did the profiling - I will post it below.

But actually observe having all unique IDs is a point of working `stack`/`unstack` (i.e. they are unique after `unstack` but are not unique after `stack`).